lib/utils.js

```diff
@@ -229,6 +230,147 @@ const format = body => {
+ */
+const references = (types) => {
+  const cts = _.cloneDeep(types);
+  const refs = [];
+  const position = {};
+
+  // Saves position of each content type
+  cts.forEach((type, index) => {
+    if (!position.hasOwnProperty(type.id)) {
+      position[type.id] = index;
+    }
+  });
+
+  // Iterates through each content type and attributes
+  cts.map(ct => {
+    ct.attributes.map((attr, index) => {
+      let inputs;
+      let length;
+
+      // Get the number of instances if repeatable
+      if (Array.isArray(attr.inputs)) {
+        inputs = attr.inputs[0];
+        length = attr.inputs.length;
+      }
+      else {
+        inputs = attr.inputs;
+      }
+
+      // Iterate through each inputs
+      Object.keys(inputs).forEach(input => {
+        // Checks if input is a reference
+        if (inputs[input].hasOwnProperty('reference') && inputs[input].reference === true) {
+          // Throw error if contentType is not defined
+          if (!inputs[input].settings.hasOwnProperty('contentType')) {
+            throw new Error('Reference must have a content type');
+          }
+
+          // Throw error if contentType is not valid
+          if (!position.hasOwnProperty(inputs[input].settings.contentType)) {
+            throw new Error(`Content Type ${inputs[input].settings.contentType} is not valid`);
+          }
+
+          // Defines and push reference object
+          refs.push({
+            type: ct.id,
+            attr: index,
+            input,
+            length,
+            ct: {
+              index: position[inputs[input].settings.contentType],
+              id: inputs[input].settings.contentType,
+            },
+          });
+        }
+      });
+    });
+  });
+
+  return refs;
+};
+
+
+/*
+ * Fills in content type data from database
+ *
+ * @param {object} types - content types object
+ * @param {string} id - id of the requested content type
+ * @param {array} refs - reference data
+ * @param {object} database - database object
+ *
+ *
+ * @returns {object} - content type object with reference data
+ */
+const fill = (types, id, refs, database) => {
+  const ct = singleItem('id', id, types);
+  if (!ct) {
+    return new Promise(resolve => {
+      resolve(ct);
+    });
+  }
+  const ref = refs.filter(reference => {
+    return ct.id === reference.type;
+  });
+
+  return Promise.map(ref, input => {
+    // Get the records from content-type
+    return database
+      .select('*')
+      .from(`content-type--${input.ct.id}`)
+      .orderBy('revision', 'DESC').then(rws => {
+        const options = [];
+        const map = {};
+
+        // Gets the lates revison and pushes unique values to options
+        rws.forEach(row => {
+          if (!map.hasOwnProperty(row.id)) {
+            map[row.id] = true;
+
+            // Checks if data is stored in the correct structure
+            if (row.value.hasOwnProperty(types[input.ct.index].identifier)) {
```

Comment:
We should make the `identifier` attribute automatically `require: save`. Then we wouldn't need to have this check.